File: app/src/main/java/cn/ifafu/ifafu/mvp/main/MainModel.java
package cn.ifafu.ifafu.mvp.main;

import android.content.Context;
import android.graphics.drawable.Drawable;

import com.alibaba.fastjson.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import cn.ifafu.ifafu.R;
import cn.ifafu.ifafu.app.Constant;
import cn.ifafu.ifafu.app.School;
import cn.ifafu.ifafu.dao.DaoSession;
import cn.ifafu.ifafu.data.entity.Holiday;
import cn.ifafu.ifafu.data.entity.Menu;
import cn.ifafu.ifafu.data.entity.User;
import cn.ifafu.ifafu.data.entity.Weather;
import cn.ifafu.ifafu.data.http.APIManager;
import cn.ifafu.ifafu.data.http.service.WeatherService;
import cn.ifafu.ifafu.data.local.DaoManager;
import cn.ifafu.ifafu.electricity.splash.ElecSplashActivity;
import cn.ifafu.ifafu.mvp.base.BaseZFModel;
import cn.ifafu.ifafu.mvp.exam.ExamActivity;
import cn.ifafu.ifafu.mvp.score.ScoreActivity;
import cn.ifafu.ifafu.mvp.syllabus.SyllabusActivity;
import cn.ifafu.ifafu.mvp.web.WebActivity;
import cn.ifafu.ifafu.util.SPUtils;
import io.reactivex.Observable;
import okhttp3.ResponseBody;

public class MainModel extends BaseZFModel implements MainContract.Model {

    private User user = repository.getUser();

    MainModel(Context context) {
        super(context);
    }

    @Override
    public Observable<List<Menu>> getMenus() {
        return Observable.fromCallable(() -> {
            List<Menu> menus = new ArrayList<>();
            menus.add(new Menu(mContext.getDrawable(R.drawable.tab_syllabus), "课程表", SyllabusActivity.class));
            menus.add(new Menu(mContext.getDrawable(R.drawable.tab_exam), "考试计划", ExamActivity.class));
            menus.add(new Menu(mContext.getDrawable(R.drawable.tab_score), "成绩查询", ScoreActivity.class));
            menus.add(new Menu(mContext.getDrawable(R.drawable.tab_web), "网页模式", WebActivity.class));
            menus.add(new Menu(mContext.getDrawable(R.drawable.ic_electricity), "电费查询", ElecSplashActivity.class));
            return menus;
        });
    }

    @Override
    public Drawable getSchoolIcon() {
        switch (user.getSchoolCode()) {
            case School.FAFU:
                return mContext.getDrawable(R.drawable.drawable_fafu2);
            case School.FAFU_JS:
                return mContext.getDrawable(R.drawable.drawable_fafu_js2);
            default:
                return mContext.getDrawable(R.drawable.drawable_ifafu2);
        }
    }

    @Override
    public String getUserName() {
        return user.getName();
    }

    @Override
    public Observable<Weather> getWeather(String cityCode) {
        return Observable.fromCallable(() -> {
            Weather weather = new Weather();
            String referer = "http://www.weather.com.cn/weather1d/" + cityCode + ".shtml";
            WeatherService service = APIManager.getWeatherAPI();

            // 获取城市名和当前温度
            String url1 = "http://d1.weather.com.cn/sk_2d/" + cityCode + ".html";
            ResponseBody body1 = service.getWeather(url1, referer).execute().body();
            String jsonStr1 = Objects.requireNonNull(body1).string();
            jsonStr1 = jsonStr1.replace("var dataSK = ", "");
            JSONObject jo1 = JSONObject.parseObject(jsonStr1);
            weather.setCityName(jo1.getString("cityname"));
            weather.setNowTemp(jo1.getInteger("temp"));
            weather.setWeather(jo1.getString("weather"));

            // 获取白天温度和晚上温度
            String url2 = "http://d1.weather.com.cn/dingzhi/" + cityCode + ".html";
            ResponseBody body2 = service.getWeather(url2, referer).execute().body();
            String jsonStr2 = Objects.requireNonNull(body2).string();
            jsonStr2 = jsonStr2.substring(jsonStr2.indexOf('=')+1, jsonStr2.indexOf(";"));
            JSONObject jo2 = JSONObject.parseObject(jsonStr2);
            jo2 = jo2.getJSONObject("weatherinfo");
            weather.setAmTemp(Integer.valueOf(jo2.getString("temp").replace("℃", "")));
            weather.setPmTemp(Integer.valueOf(jo2.getString("tempn").replace("℃", "")));

            return weather;
        });
    }

    @Override
    public List<Holiday> getHoliday() {
        List<Holiday> holidays = new ArrayList<>();
        holidays.add(new Holiday("教师节", "2019-09-10"));
        holidays.add(new Holiday("中秋节", "2019-09-13"));
        holidays.add(new Holiday("国庆节", "2019-10-01"));
        holidays.add(new Holiday("圣诞节", "2019-12-25"));
        holidays.add(new Holiday("元旦", "2020-01-01"));
        holidays.add(new Holiday("春节", "2020-01-25"));
        holidays.add(new Holiday("清明节", "2020-04-04"));
        holidays.add(new Holiday("劳动节", "2020-05-01"));
        return holidays;
    }

    @Override
    public void clearAllDate() {
        SPUtils.get(Constant.SP_SETTING).clear();
        SPUtils.get(Constant.SP_USER_INFO).clear();
        DaoSession daoSession = DaoManager.getInstance().getDaoSession();
        daoSession.getScoreDao().deleteAll();
        daoSession.getCourseDao().deleteAll();
        daoSession.getExamDao().deleteAll();
        daoSession.getUserDao().deleteAll();
    }
}

File: app/src/main/java/cn/ifafu/ifafu/mvp/main/MainModel.kt
package cn.ifafu.ifafu.mvp.main

import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import androidx.annotation.DrawableRes
import cn.ifafu.ifafu.R
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.app.School
import cn.ifafu.ifafu.data.entity.Holiday
import cn.ifafu.ifafu.data.entity.Menu
import cn.ifafu.ifafu.data.entity.Weather
import cn.ifafu.ifafu.data.http.APIManager
import cn.ifafu.ifafu.data.http.service.WeatherService
import cn.ifafu.ifafu.electricity.splash.ElecSplashActivity
import cn.ifafu.ifafu.mvp.base.BaseZFModel
import cn.ifafu.ifafu.mvp.exam.ExamActivity
import cn.ifafu.ifafu.mvp.main.MainContract.Model
import cn.ifafu.ifafu.mvp.score.ScoreActivity
import cn.ifafu.ifafu.mvp.syllabus.SyllabusActivity
import cn.ifafu.ifafu.mvp.syllabus.SyllabusModel
import cn.ifafu.ifafu.mvp.web.WebActivity
import com.alibaba.fastjson.JSONObject
import io.reactivex.Observable
import java.util.*

class MainModel(context: Context?) : BaseZFModel(context), Model {

    override fun getMenus(): Observable<List<Menu>> {
        return Observable.fromCallable {
            listOf(Menu(drawable(R.drawable.tab_syllabus), "课程表", intent(SyllabusActivity::class.java)),
                    Menu(drawable(R.drawable.tab_exam), "考试计划", intent(ExamActivity::class.java)),
                    Menu(drawable(R.drawable.tab_score), "成绩查询", intent(ScoreActivity::class.java)),
                    Menu(drawable(R.drawable.tab_web), "网页模式", intent(WebActivity::class.java)),
                    Menu(drawable(R.drawable.tab_electricity), "电费查询", intent(ElecSplashActivity::class.java)),
                    Menu(drawable(R.drawable.tab_repair), "报修服务", intent(WebActivity::class.java).apply {
                        putExtra("title", "报修服务")
                        putExtra("url", Constant.REPAIR_URL)
                    })
            )
        }
    }

    private fun drawable(@DrawableRes id: Int): Drawable? = mContext.getDrawable(id)
    private fun intent(cls: Class<*>): Intent = Intent(mContext, cls)

    override fun getSchoolIcon(): Drawable? {
        return when (repository.user.schoolCode) {
            School.FAFU -> drawable(R.drawable.drawable_fafu_b_white)
            School.FAFU_JS -> drawable(R.drawable.drawable_fafu_js_white)
            else -> drawable(R.mipmap.ic_launcher_round)
        }
    }

    override fun getUserName(): String {
        return repository.user.name
    }

    override fun getWeather(cityCode: String): Observable<Weather> {
        return Observable.fromCallable {
            val weather = Weather()
            val referer = "http://www.weather.com.cn/weather1d/$cityCode.shtml"
            val service: WeatherService = APIManager.getWeatherAPI()

            // 获取城市名和当前温度
            val url1 = "http://d1.weather.com.cn/sk_2d/$cityCode.html"
            val body1 = service.getWeather(url1, referer).execute().body()
            var jsonStr1: String = Objects.requireNonNull(body1)!!.string()
            jsonStr1 = jsonStr1.replace("var dataSK = ", "")
            val jo1: JSONObject = JSONObject.parseObject(jsonStr1)
            weather.cityName = jo1.getString("cityname")
            weather.nowTemp = jo1.getInteger("temp")
            weather.weather = jo1.getString("weather")

            // 获取白天温度和晚上温度
            val url2 = "http://d1.weather.com.cn/dingzhi/$cityCode.html"
            val body2 = service.getWeather(url2, referer).execute().body()
            var jsonStr2: String = Objects.requireNonNull(body2)!!.string()
            jsonStr2 = jsonStr2.substring(jsonStr2.indexOf('=') + 1, jsonStr2.indexOf(";"))
            var jo2: JSONObject = JSONObject.parseObject(jsonStr2)
            jo2 = jo2.getJSONObject("weatherinfo")
            weather.amTemp = Integer.valueOf(jo2.getString("temp").replace("℃", ""))
            weather.pmTemp = Integer.valueOf(jo2.getString("tempn").replace("℃", ""))
            weather
        }
    }

    override fun getHoliday(): List<Holiday> {
        return SyllabusModel(mContext).holidays
    }

    override fun clearAllDate() {
        repository.clearAllData()
    }
}